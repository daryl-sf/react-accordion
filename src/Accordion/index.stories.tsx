import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import Accordion from ".";

export default {
  component: Accordion,
} as ComponentMeta<typeof Accordion>;

const AccordionTemplate: ComponentStory<typeof Accordion> = () => (
  <div style={{ width: "600px" }}>
    <Accordion
      id="first"
      renderSummary={({ onClick, expanded }) => (
        <div
          style={{
            display: "flex",
            justifyContent: "space-between",
            marginBottom: "6px",
          }}
        >
          summary{" "}
          <button
            type="button"
            onClick={onClick}
            style={{ all: "unset", color: "#ff5050", cursor: "pointer" }}
          >
            {expanded ? "show less" : "show more"}
          </button>
        </div>
      )}
      renderDetail={
        <div>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean
          laoreet erat at mi laoreet, vitae rutrum tellus suscipit. Lorem ipsum
          dolor sit amet, consectetur adipiscing elit. Maecenas a magna
          condimentum, consectetur ante vel, imperdiet urna. Nulla dictum
          bibendum augue a blandit. Donec ullamcorper purus eget ornare
          bibendum. Nunc auctor arcu a dui vestibulum, vitae rutrum tellus
          egestas. Ut pharetra tellus id est facilisis, rhoncus facilisis odio
          bibendum. Vivamus dignissim enim elit, quis condimentum ligula gravida
          id. Nulla varius suscipit dolor. Fusce maximus tellus sed tellus
          convallis, vitae feugiat enim ullamcorper. Donec laoreet finibus urna
          sit amet condimentum. Maecenas efficitur aliquam libero sit amet
          malesuada.
        </div>
      }
    />
    <br />
    <Accordion
      id="second"
      initExpanded
      renderSummary={({ onClick, expanded }) => (
        <div
          style={{
            display: "flex",
            justifyContent: "space-between",
            marginBottom: "6px",
          }}
        >
          Open me to get more info about how this works
          <button
            type="button"
            onClick={onClick}
            style={{ all: "unset", color: "#ff5050", cursor: "pointer" }}
          >
            {expanded ? "show less" : "show more"}
          </button>
        </div>
      )}
      renderDetail={
        <div>
          <p>
            This accordion transitions <code>max-height</code> from 0 to x and
            vice versa when <code>expanded</code> state changes.
          </p>
          <p>
            The component is designed to handle dynamic content that can vary in
            height. <code>heightRef</code> is a div with{" "}
            <code>overflow:hidden</code>.
          </p>
          <p>
            This is used as a ref so we know what height the content is and
            therefore what value to set <code>max-height</code> to.
          </p>
        </div>
      }
    />
  </div>
);

export const AccordionStory = AccordionTemplate.bind({});
