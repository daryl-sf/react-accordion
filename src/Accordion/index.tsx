import React, { ReactNode, useState, useRef, useEffect } from "react";
import './styles.css'

// This accordion transitions max-height from 0 to x and vice versa when expanded state changes.
// The component is designed to handle dynamic content that can vary in height. heightRef is a div with overflow:hidden.
// This is used as a ref so we know what height the content is and therefore what value to set max-height to.

export type IAccordionProps = {
  id: string;
  initExpanded?: boolean;
  renderSummary: (props: {
    onClick: () => void;
    expanded: boolean;
  }) => JSX.Element;
  renderDetail: ReactNode;
  className?: string;
}

const Accordion = ({
  id,
  initExpanded = false,
  renderSummary,
  renderDetail,
  className = "",
}: IAccordionProps) => {
  const [expanded, setExpanded] = useState(initExpanded);
  const [maxHeight, setMaxHeight] = useState(0);
  const heightRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    setMaxHeight(heightRef.current?.clientHeight || 0);
  }, [heightRef]);

  return (
    <div className={`container ${className}`} id={id}>
      {renderSummary({ onClick: () => setExpanded(!expanded), expanded })}
      <div
        className={`detail ${expanded ? expanded : '' }`}
        style={{
          maxHeight: expanded ? maxHeight : 0,
        }}
      >
        <div className="heightRef" ref={heightRef}>
          {renderDetail}
        </div>
      </div>
    </div>
  );
};

export default Accordion;