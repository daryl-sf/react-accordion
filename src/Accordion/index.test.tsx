import React from "react";
import { render } from "@testing-library/react";

import Accordion from ".";

describe("Accordion", () => {
  test("renders the Accordion component without error", () => {
    const { asFragment } = render(
      <Accordion
        id="testAccordion"
        renderDetail={<div>This is an accordion</div>}
        renderSummary={({ onClick, expanded }) => (
          <div>This is an accordion. <button type="button" onClick={onClick}>Click me to {expanded ? "collapse" : "expand"}</button></div>
        )}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe("initExpanded = true", () => {
    test("renders accordion with maxHeight != 0", () => {
      const { asFragment } = render(
        <Accordion
          id="testAccordion"
          initExpanded={true}
          renderDetail={<div style={{ height: "30px"}}>This is an accordion</div>}
          renderSummary={({ onClick, expanded }) => (
            <div>This is an accordion. <button type="button" onClick={onClick}>Click me to {expanded ? "collapse" : "expand"}</button></div>
          )}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    })
  });
});
