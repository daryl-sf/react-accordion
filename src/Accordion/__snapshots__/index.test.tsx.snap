// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion initExpanded = true renders accordion with maxHeight != 0 1`] = `
<DocumentFragment>
  <div
    class="container "
    id="testAccordion"
  >
    <div>
      This is an accordion. 
      <button
        type="button"
      >
        Click me to collapse
      </button>
    </div>
    <div
      class="detail true"
      style="max-height: 5px;"
    >
      <div
        class="heightRef"
      >
        <div
          style="height: 30px;"
        >
          This is an accordion
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Accordion renders the Accordion component without error 1`] = `
<DocumentFragment>
  <div
    class="container "
    id="testAccordion"
  >
    <div>
      This is an accordion. 
      <button
        type="button"
      >
        Click me to expand
      </button>
    </div>
    <div
      class="detail "
      style="max-height: 0;"
    >
      <div
        class="heightRef"
      >
        <div>
          This is an accordion
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
